File: Practice/20/c++/vine/vine/vine.cpp

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Drink
{
public:
	Drink()
	{
		name = ' ';
		price = 0;
		volume = 1;
		perv = price / volume;
		k = 0; 
	}

	Drink(string name, int price, int volume)
	{
		this->name = name;
		this->price = price;
		this->volume = volume;
		this->perv = price / volume;
	}

	Drink(const Drink &other)
	{
		this->name = other.name;
		this->price = other.price;
		this->volume = other.volume;
		this->k = other.k;
		this->perv = other.perv;
	}

	int getPrice()
	{
		return price;
	}

	int getK()
	{
		return k;
	}

	int getVolume()
	{
		return volume;
	}

	string getName()
	{
		return name;
	}

	void addK(int c)
	{
		k = c;
	}

	bool operator <(const Drink &other)
	{
		return (this->perv < other.perv);
	}

	bool operator >(const Drink& other)
	{
		return (this->perv > other.perv);
	}

double perv;
private:
	string name;
	int price;
	int volume;
	
	int k;
};

int main(){
	setlocale(LC_ALL, "rus");
	
	int buget;
	int assorti;
	string name;
	int price;
	int volume;
	int temp;
	int sumvol = 0;
	int minprice = 1000000001;

	cout << "Введите количество денег\n";
	cin >> buget;

	cout << "Введите количество разновидностей напитков\n";
	cin >> assorti;

	vector <Drink> drinks(assorti);

	cout << "Введите наименование напитка, его цену и его объем через пробел, причем каждый новый напиток с новой строки.\n";
	for (int i = 0; i < assorti; i++) {
		cin >> name;
		cin >> price;
		cin >> volume;
		drinks[i] = Drink(name, price, volume);
		if (minprice > price) minprice = price;
	}
	if (minprice > buget) cout << -1;
	else {
		sort(drinks.begin(), drinks.end(), [](Drink a, Drink b) {return a < b; });

		for (int i = 0; i < assorti; i++) {
			drinks[i].addK(buget / drinks[i].getPrice());
			buget %= drinks[i].getPrice();
		}
		cout << "<Название напитка> <количество бутылок>:\n";
		for (int i = 0; i < assorti; i++) {
			temp = drinks[i].getK();
			if (temp) {
				cout << drinks[i].getName() << " " << temp << "\n";
				sumvol += drinks[i].getVolume() * temp;
			}
		}
		cout << "Суммарный объем: " << sumvol << "\n";
		cout << "Остаток денег: " << buget;
	}
}

File: Practice/20/py/vine/vine/vine.py
class Drink:
    def __init__(self, name, price, volume):
        self.name = name
        self.price = price
        self.volume = volume
        self.perv = price/volume
        self.k = 0

    def getPrice(self):
        return self.price

    def getK(self):
        return self.k

    def getName(self):
        return self.name

    def getVolume(self):
        return self.volume
    
    def addK(self, a):
        self.k = a

buget = int(input("Введите количество денег\n"))
assorti = int(input("Введите количество разновидностей напитков\n"))
print("Введите наименование напитка, его цену и его объем через пробел, причем каждый новый напиток с новой строки.\n")
drinks = []
minprice = 1000000001
sumvol = 0
for i in range (assorti):
    name, price, volume= input().split(" ")
    price = int(price)
    volume = int(volume)
    drinks.append(Drink(name, price, volume))
    if(minprice > price): minprice = price
if(minprice > buget): print(-1)
else:
    sorted(drinks, key = lambda drink: drink.perv)
    for i in range (assorti):
        drinks[i].addK(buget // drinks[i].getPrice())
        buget %= drinks[i].getPrice();
    print("<Название напитка> <количество бутылок>:\n")
    for i in range (assorti):
        temp = drinks[i].getK()
        if (temp):
            print(str(drinks[i].getName()) + " " + str(temp) + "\n")
            sumvol += drinks[i].getVolume() * temp
    print("Суммарный объем: " + str(sumvol) + "\n")
    print("Остаток денег: " + str(buget))